pkg/logql/filter.go

```diff
@@ -0,0 +1,287 @@
+package logql
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"regexp/syntax"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+// LineFilter is a interface to filter log lines.
+type LineFilter interface {
+	Filter(line []byte) bool
+}
+
+// LineFilterFunc is a syntax sugar for creating line filter from a function
+type LineFilterFunc func(line []byte) bool
+
+func (f LineFilterFunc) Filter(line []byte) bool {
+	return f(line)
+}
+
+type notFilter struct {
+	LineFilter
+}
+
+func (n notFilter) Filter(line []byte) bool {
+	return !n.LineFilter.Filter(line)
+}
+
+// newNotFilter creates a new filter which matches only if the base filter doesn't match.
+// If the base filter is a `or` it will recursively simplify with `and` operations.
+func newNotFilter(base LineFilter) LineFilter {
+	// not(a|b) = not(a) and not(b)
+	if or, ok := base.(orFilter); ok {
+		return newAndFilter(newNotFilter(or.left), newNotFilter(or.right))
+	}
+	return notFilter{LineFilter: base}
+}
+
+type andFilter struct {
+	left  LineFilter
+	right LineFilter
+}
+
+// newAndFilter creates a new filter which matches only if left and right matches.
+func newAndFilter(left LineFilter, right LineFilter) LineFilter {
+	return andFilter{
+		left:  left,
+		right: right,
+	}
+}
+
+func (a andFilter) Filter(line []byte) bool {
+	return a.left.Filter(line) && a.right.Filter(line)
+}
+
+type orFilter struct {
+	left  LineFilter
+	right LineFilter
+}
```

Comment:
```suggestion
type orFilter []LineFilter
```

Have you considered this alternative? Then `newOrFilter` can then work as `chainOrFilter`, and also trivially merge two `orFilters` together, and filtering simply checks if any of them matches. Same for `andFilter`.

Reply:
ho that's a good idea, I'll give a go.

Reply:
because of how I deal with not or I can't really do that easily. Let's keep this for later.